### ISSUE TITLE
Wrong epic link if task has changed epic

### ISSUE THREAD

**Issue by dakimaz** [solidify/jira-azuredevops-migrator#356]
If a given task has changed epic in its history the import will throw an error:

> [E][09:07:09] [Microsoft.TeamFoundation.WorkItemTracking.Client.WorkItemLinkValidationException] TF201066: You cannot add a Child link to work item 13558 because a work item can have only one Parent link.. Link Source: 13575, Target: 13558 in 'WSWD-372', rev 1 will be skipped.
> [Microsoft.TeamFoundation.WorkItemTracking.Client.WorkItemLinkValidationException] Microsoft.TeamFoundation.WorkItemTracking.Client.WorkItemLinkValidationException: TF201066: You cannot add a Child link to work item 13558 because a work item can have only one Parent link. ---> System.Web.Services.Protocols.SoapException: TF201036: You cannot add a Child link between work items 13575 and 13558 because a work item can have only one Parent link.
>    at Microsoft.TeamFoundation.WorkItemTracking.Proxy.RetryHandler.HandleSoapException(SoapException se)
>    at Microsoft.TeamFoundation.WorkItemTracking.Proxy.WorkItemServer.Update(String requestId, XmlElement package, XmlElement& result, MetadataTableHaveEntry[] metadataHave, String& dbStamp, IMetadataRowSets& metadata)
>    at Microsoft.TeamFoundation.WorkItemTracking.Client.WorkItemStore.SendUpdatePackage(XmlElement package, XmlElement& result, Boolean bulk)
>    --- End of inner exception stack trace ---
>    at Microsoft.TeamFoundation.WorkItemTracking.Client.WorkItem.Save(SaveFlags saveFlags)
>    at WorkItemImport.Agent.SaveWorkItem(WiRevision rev, WorkItem newWorkItem): 
>    at Microsoft.TeamFoundation.WorkItemTracking.Client.WorkItem.Save(SaveFlags saveFlags)
>    at WorkItemImport.Agent.SaveWorkItem(WiRevision rev, WorkItem newWorkItem)

When parsing the tasks it sees that the task in question already has an epic assigned and it fails.

It seems that instead of replacing the epic that is linked, it is simply trying to add another epic link. The result is a missing revision entry and incorrect epic link.

Could it be that I am missing an entry in the link-map? I will admit that I am struggling with understanding if, for example, it should be:

```
      {
        "source": "blocks",
        "target": "System.LinkTypes.Dependency-Reverse"
      },
      {
        "source": "blocked by",
        "target": "System.LinkTypes.Dependency-Forward"
      },
```

Or if this is sufficient:

```
      {
        "source": "Blocked",
        "target": "System.LinkTypes.Dependency"
      }
```
I.e; if the tool can automatically figure out that `blocks -> Reverse` and `blocked by -> Forward`

What confuses me even further is, that all samples are using:

```
      {
        "source": "Epic",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
```

But https://confluence.atlassian.com/adminjiraserver/files/938847862/938847863/1/1507216097370/issue_linking.png shows that Epics can have an "inbound" and "outbound" type, so should it instead be:

```
      {
        "source": "belongs to Epic",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "is the Epic for",
        "target": "System.LinkTypes.Hierarchy-Forward"
      },
```
Any help would be appreciated.

Best Regards

AB#4953

**Comment by bggo** [comment]
Guys!

Any thoughts on that! Same error here, its only happening with histories that have changed epic!

Regards

**Comment by Alexander-Hjelm** [comment]
No immediate thoughts. We will possibly address it in an upcoming release, since we have introduced automated testing with my latest PR: 

https://github.com/solidify/jira-azuredevops-migrator/pull/458

@bggo  if possible, could you describe more about under what circumstances the bug appears? For any task/issue that has changed their parent epic?

**Comment by Alexander-Hjelm** [comment]
Tested for an Agile project and verified that the issue has been resolved. See attached images for results.

@dakimaz @bggo, I will close this issue within a week or so if I don't hear otherwise. Feel free to reopen or create a new issue and tag me if the problem persists. :)

![image](https://user-images.githubusercontent.com/10683896/233846801-ca038c5c-072f-4ee7-8e01-e8836f5a8110.png)
![image](https://user-images.githubusercontent.com/10683896/233846820-1b606635-603b-4569-8d66-e1871d00a4c2.png)

**Comment by DavidRadoorHummel** [comment]
This is great!
Thanks Alexander! We will check it out :)

Best regards
David

**Comment by Alexander-Hjelm** [comment]
Closing since issue already seems to have been resolved
